File: goldencheck/profilers/sequence_detection.py
"""Sequence gap detection profiler — detects gaps in sequential integer columns."""
from __future__ import annotations
import polars as pl
from goldencheck.models.finding import Finding, Severity
from goldencheck.profilers.base import BaseProfiler

INTEGER_DTYPES = (
    pl.Int8, pl.Int16, pl.Int32, pl.Int64,
    pl.UInt8, pl.UInt16, pl.UInt32, pl.UInt64,
)

# Minimum fraction of consecutive diffs == 1 to consider column sequential.
# We use this threshold on columns where the values increment exactly by 1 most of the time.
# For columns with gaps (diffs > 1) we apply a looser check: is the column sorted ascending
# and are >=90% of diffs positive?
SEQUENTIAL_THRESHOLD = 0.90


class SequenceDetectionProfiler(BaseProfiler):
    def profile(self, df: pl.DataFrame, column: str, *, context: dict | None = None) -> list[Finding]:
        findings: list[Finding] = []
        col = df[column]

        if col.dtype not in INTEGER_DTYPES:
            return findings

        non_null = col.drop_nulls()
        total = len(non_null)
        if total < 2:
            return findings

        # Compute consecutive differences
        diffs = non_null.diff().drop_nulls()
        n_diffs = len(diffs)
        if n_diffs == 0:
            return findings

        # A column is considered "sequential" when:
        #   - >=90% of diffs are exactly 1 (tight sequential), OR
        #   - >=90% of diffs are positive AND the values are sorted ascending
        #     (sequential with gaps — still clearly an ID/sequence column)
        unit_diffs = int((diffs == 1).sum())
        positive_diffs = int((diffs > 0).sum())
        sequential_ratio = unit_diffs / n_diffs
        positive_ratio = positive_diffs / n_diffs

        is_tight_sequential = sequential_ratio >= SEQUENTIAL_THRESHOLD
        is_sorted_sequential = (positive_ratio >= SEQUENTIAL_THRESHOLD) and non_null.is_sorted()

        if not (is_tight_sequential or is_sorted_sequential):
            # Not sequential — skip
            return findings

        # Column is sequential — find gaps
        col_min = int(non_null.min())
        col_max = int(non_null.max())
        expected_count = col_max - col_min + 1

        if expected_count <= total:
            # No gaps
            return findings

        # Find the actual gaps
        full_range = pl.Series("expected", range(col_min, col_max + 1))
        present = non_null.unique().sort()
        gaps = full_range.filter(~full_range.is_in(present))
        gap_count = len(gaps)

        sample_gaps = gaps.head(10).to_list()
        findings.append(Finding(
            severity=Severity.WARNING,
            column=column,
            check="sequence_detection",
            message=(
                f"Sequence gap detected in column '{column}': "
                f"{gap_count} missing value(s) in range [{col_min}, {col_max}]. "
                f"Gap positions (sample): {sample_gaps}"
            ),
            affected_rows=gap_count,
            sample_values=[str(v) for v in sample_gaps],
            suggestion="Investigate whether the missing sequence numbers indicate deleted or skipped records",
            confidence=0.7,
        ))

        return findings

File: packages/goldencheck-js/src/core/profilers/sequence-detection.ts
/**
 * Sequence gap detection profiler — detects gaps in sequential integer columns.
 * Port of goldencheck/profilers/sequence_detection.py.
 */

import type { TabularData } from "../data.js";
import { type Finding, Severity, makeFinding } from "../types.js";
import type { Profiler } from "./base.js";

const SEQUENTIAL_THRESHOLD = 0.9;

export class SequenceDetectionProfiler implements Profiler {
  profile(data: TabularData, column: string): Finding[] {
    const findings: Finding[] = [];

    const dt = data.dtype(column);
    if (dt !== "integer") return findings;

    const nums = data.numericValues(column).filter(Number.isInteger);
    if (nums.length < 2) return findings;

    // Compute consecutive differences
    const diffs: number[] = [];
    for (let i = 1; i < nums.length; i++) {
      diffs.push(nums[i]! - nums[i - 1]!);
    }

    if (diffs.length === 0) return findings;

    const unitDiffs = diffs.filter((d) => d === 1).length;
    const positiveDiffs = diffs.filter((d) => d > 0).length;
    const sequentialRatio = unitDiffs / diffs.length;
    const positiveRatio = positiveDiffs / diffs.length;

    // Check if sorted ascending
    const isSorted = nums.every((v, i) => i === 0 || v >= nums[i - 1]!);

    const isTightSequential = sequentialRatio >= SEQUENTIAL_THRESHOLD;
    const isSortedSequential = positiveRatio >= SEQUENTIAL_THRESHOLD && isSorted;

    if (!isTightSequential && !isSortedSequential) return findings;

    // Find gaps
    const colMin = Math.min(...nums);
    const colMax = Math.max(...nums);
    const expectedCount = colMax - colMin + 1;

    if (expectedCount <= nums.length) return findings;

    const present = new Set(nums);
    const gaps: number[] = [];
    for (let v = colMin; v <= colMax; v++) {
      if (!present.has(v)) gaps.push(v);
    }

    const sampleGaps = gaps.slice(0, 10);
    findings.push(
      makeFinding({
        severity: Severity.WARNING,
        column,
        check: "sequence_detection",
        message: `Sequence gap detected in column '${column}': ${gaps.length} missing value(s) in range [${colMin}, ${colMax}]. Gap positions (sample): ${JSON.stringify(sampleGaps)}`,
        affectedRows: gaps.length,
        sampleValues: sampleGaps.map(String),
        suggestion: "Investigate whether the missing sequence numbers indicate deleted or skipped records",
        confidence: 0.7,
      }),
    );

    return findings;
  }
}
